### PR TITLE
Update client.rs

### PR DESCRIPTION
**Pull request by dev4u**
Fix the issue that the connection creation fails when the network changes

### CODE REVIEW

**Comment by zonyitoo** [comment]
This will create a new UdpSocket for every accepted connections. Why do you think that it is the root cause?

**Comment by dev4u** [comment]
因为手机的网络环境发生变化的时候，之前建立的nat映射失效了。不rebind的话发不了数据。
当然，我也承认提交的代码比较粗糙，或许可以先检测网络环境是否有发生变化，再决定要不要rebind，这比较合乎该有的效果。

**Comment by madeye** [comment]
You should only do this when the `connect()` failed: https://github.com/shadowsocks/qtun/blob/master/src/client.rs#L116

**Comment by dev4u** [comment]
@zonyitoo 你替我看看这样修改还有没有问题？

**Comment by dev4u** [comment]
> You should only do this when the `connect()` failed: https://github.com/shadowsocks/qtun/blob/master/src/client.rs#L116

已经按照你的建议做了修改。

**Comment by madeye** [comment]
Can you confirm your fix works now?

**Comment by dev4u** [comment]
> Can you confirm your fix works now?

是的，这两天一直在我的手机移动网络下运行着。

File: src/client.rs [hunk 1]
```diff
@@ -121,7 +115,19 @@ async fn transfer(
...
-        .map_err(|e| anyhow!("failed to connect: {}", e))?;
+        .map_err(|e| {
+            if e == ConnectionError::TimedOut {
+                let socket = if cfg!(target_os = "windows") {
+                    std::net::UdpSocket::bind("0.0.0.0:0").unwrap()
+                } else {
+                    std::net::UdpSocket::bind("[::]:0").unwrap()
+                };
+                endpoint.rebind(socket)?;
+                Ok(())
```
Comment (madeye) [inline src/client.rs]: You still need to return the error here.

File: src/client.rs [hunk 2]
```diff
@@ -121,7 +115,19 @@ async fn transfer(
...
         .await
-        .map_err(|e| anyhow!("failed to connect: {}", e))?;
+        .map_err(|e| {
+            if e == ConnectionError::TimedOut {
+                let socket = if cfg!(target_os = "windows") {
+                    std::net::UdpSocket::bind("0.0.0.0:0").unwrap()
+                } else {
+                    std::net::UdpSocket::bind("[::]:0").unwrap()
+                };
+                endpoint.rebind(socket)?;
```
Comment (madeye) [inline src/client.rs]: It'd be better log the rebind event.

**Comment by madeye** [comment]
https://github.com/shadowsocks/qtun/runs/5578803045?check_suite_focus=true

**Pull request merged**